--- modules/nodereference/nodereference.orig.module	2009-04-06 20:14:25.000000000 -0400
+++ modules/nodereference/nodereference.module	2009-04-06 20:19:09.000000000 -0400
@@ -411,6 +411,7 @@ function nodereference_widget_settings($
     case 'form':
       $form = array();
       $match = isset($widget['autocomplete_match']) ? $widget['autocomplete_match'] : 'contains';
+      $size = (isset($widget['size']) && is_numeric($widget['size'])) ? $widget['size'] : 60;
       if ($widget['type'] == 'nodereference_autocomplete') {
         $form['autocomplete_match'] = array(
           '#type' => 'select',
@@ -422,14 +423,22 @@ function nodereference_widget_settings($
           ),
           '#description' => t('Select the method used to collect autocomplete suggestions. Note that <em>Contains</em> can cause performance issues on sites with thousands of nodes.'),
         );
+        $form['size'] = array(
+          '#type' => 'textfield',
+          '#title' => t('Size of textfield'),
+          '#default_value' => $size,
+          '#element_validate' => array('_element_validate_integer_positive'),
+          '#required' => TRUE,
+        );
       }
       else {
         $form['autocomplete_match'] = array('#type' => 'hidden', '#value' => $match);
+        $form['size'] = array('#type' => 'hidden', '#value' => $size);
       }
       return $form;
 
     case 'save':
-      return array('autocomplete_match');
+      return array('autocomplete_match', 'size');
   }
 }
 
