Index: content_profile.module
===================================================================
RCS file: /cvs/drupal/contributions/modules/content_profile/Attic/content_profile.module,v
retrieving revision 1.1.2.32
diff -u -p -r1.1.2.32 content_profile.module
--- content_profile.module	7 Jan 2009 10:58:05 -0000	1.1.2.32
+++ content_profile.module	1 Apr 2009 05:22:16 -0000
@@ -42,7 +42,7 @@ function content_profile_menu() {
     );
   }
   foreach (content_profile_get_types('names') as $type => $type_name) {
-    $items['user/%user/profile/'. $type] = array(
+    $items['user/%user/'. $type] = array(
       'title callback' => 'check_plain',
       'title' => drupal_ucfirst($type_name),
       'page callback' => 'content_profile_page_edit',
